### ISSUE TITLE
`pub` with `macro` 2.0 breaks docs

### ISSUE THREAD

**Issue by valff** [rust-lang/rust#45266]
Using `pub` qualifier with declarative macros 2.0 breaks the `missing_docs` lint for private items, as well as exposes them with rustdoc.

The following code shouldn't produce warnings ([playground](https://play.rust-lang.org/?gist=d7a1a84c4ba8709beb946aacc4bd2238&version=nightly)):
```rust
//! ...
#![feature(decl_macro)]
#![warn(missing_docs)]

/// ...
pub mod foo {
  // Try to comment out the macro, and warnings will be disappeared.
  pub macro bar() { println!("bar"); }
  
  #[allow(dead_code)]
  fn baz() {}
}

fn main() {}
```

`rustc --version --verbose`:
```
rustc 1.22.0-nightly (dcbbfb6e8 2017-10-12)
binary: rustc
commit-hash: dcbbfb6e807fdff9c9ba80073bb755f9d9d95e31
commit-date: 2017-10-12
host: x86_64-unknown-linux-gnu
release: 1.22.0-nightly
LLVM version: 4.0
```

**Comment by petrochenkov** [comment]
The reason is that everything reachable from a public macro 2.0 is currently very pessimistically marked as reachable to the outer world (this affects linking in the first place, but also lints like missing docs, missing `Copy`/`Debug` impls, etc).

The solution could be to mark as reachable only items actually leaked by the macro (this can be hard), or to introduce one more [gradation of reachability](https://github.com/rust-lang/rust/blob/master/src/librustc/middle/privacy.rs#L23-L33) and suppress lints for it.

**Comment by matthewjasper** [comment]
This appears to have been fixed by #63114

**Comment by workingjubilee** [comment]
Appears to still emit warnings in playground as of Now, not fixed?

**Comment by DrMeepster** [comment]
Right now, it works without errors

**Comment by valff** [comment]
It's fixed now, thanks! The current warning is legitimate and unrelated to the issue.
